torchdata/dataloader2/graph/utils.py

```diff
@@ -34,6 +30,33 @@ def helper(g) -> None:  # pyre-ignore
     return dps
 
 
+def list_dps(graph: DataPipeGraph, exclude_dps: Optional[Union[DataPipe, List[DataPipe]]] = None) -> List[DataPipe]:
+    r"""
+    Given the graph of DataPipe generated by ``traverse_dps`` function,
+    return DataPipe instances excluding the provided DataPipe and its prior DataPipe graph.
+    """
+    dps: List[DataPipe] = []
```

Comment:
Are you ok to have 2 entries of the same datapipe here?
I'm talking about:
```
B, C = A.fork(2)
D = B.concat(C).map()
```
Your solution will end up listing  datapipes twice.

Reply:
I have a `cache` object to prevent the same `DataPipe` in the list for multiple times.